<table align="center" width=100%>
    <tr>
        <td width="10%">
            <img src="in_class.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                    <b> Inclass-Lab  <br>(Day 3)
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[Large Sample Test](#z)**
2. **[Small Sample Test](#t)**
3. **[Z Proportion Test](#prop)**

**Import the required libraries**

In [1]:
# type your code here
# import 'pandas' 
import pandas as pd 

# import 'numpy' 
import numpy as np

# import subpackage of matplotlib
import matplotlib.pyplot as plt

# import 'seaborn'
import seaborn as sns

# to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')

# import statsmodels
import statsmodels.api as sm

# import 'stats' package from scipy library
from scipy import stats

# import statistics to perform statistical computations
import statistics

# import a library to perform Z-test
from statsmodels.stats import weightstats as stests

### Let's begin with some hands-on practice exercises

<a id = "z"> </a>
## 1. Large Sample Test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>1. The company MangoFun that produces mango juice claims that the protein content in their juice is more than the juice produced by their competitor company FruitMix. Protein content in 50 boxes of MangoFun juice and 80 boxes of FruitMix juice is collected from a normal population which have an average protein content as 0.4 g and 0.35 g. The two samples have a standard deviation of 0.08 g for MangoFun juice and 0.05 g for FruitMix juice. Test the claim using a critical value method at a 1% level of significance.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:

time =  [2.51, 2.1, 2.18, 2.61, 1.9, 1.7, 1.95, 2.1, 1.58, 2.5, 2.19, 1.85, 1.35, 2.4, 2.22, 2.15, 1.82, 1.62, 1.26, 
         2.2, 2.35, 2.5, 2.18, 1.58, 2.24, 1.84, 2.31, 2.24, 1.75, 1.41, 1.57, 2.4]

In [3]:
# type your code here
time = [2.51, 2.1, 2.18, 2.61, 1.9, 1.7, 1.95, 2.1, 1.58, 2.5, 2.19, 1.85,
        1.35, 2.4, 2.22, 2.15, 1.82, 1.62, 1.26, 2.2, 2.35, 2.5,
        2.18, 1.58, 2.24, 1.84, 2.31, 2.24, 1.75, 1.41, 1.57, 2.4]

In [2]:
#Constructing the Hypothesis
# H0: mu1-mu2=0
# H1: mu1-mu2>=0
# It is right tailed test

n1=50
n2=80
mean1=0.4
mean2=0.35
std1=0.08
std2=0.05
null=0

# Finding the critical value
alpha=1-0.99
z_critical=stats.norm.isf(alpha)
print('z critical:', z_critical)
# Calculate the test stastic
num=mean1-mean2-null
den=(std1**2/n1)+(std2**2/n2)
test_stats=num/np.sqrt(den)
print('Test statistics', test_stats)
p_value=stats.norm.sf(test_stats)
print('Pvalue:',p_value)
if (test_stats>z_critical):
    print('Reject null hypothesis')
else:
    print('Failed to reject null hypothesis')
if (p_value<=alpha):
    print('Reject the null hypothesis')
else:
    print('Failed to reject the null hypothesis')

z critical: 2.3263478740408408
Test statistics 3.96214425875164
Pvalue: 3.7139817357365585e-05
Reject null hypothesis
Reject the null hypothesis


In [9]:
stats.norm.interval(0.99, loc=mean1-mean2-null, scale=np.sqrt(den))

(0.017494503792241142, 0.08250549620775895)

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>2. The technician wants to test whether the price of a phablet is not the same as the price of tablet. For the study, the technician collects the data of 50 gadgets of each type. Check whether the data satisfies the conditions to use the two sample Z-test. (Use 5% level of significance)</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Consider the dataset given in the CSV file `electronic.csv`. 

In [15]:
# type your code here
elec=pd.read_csv('electronic.csv')
elec.head()

Screensize (inch)  Price ($)  Weight (g)     Type
0                6.5        245         252  Phablet
1                6.3        210         252  Phablet
2                6.1        224         234  Phablet
3                6.0        217         212  Phablet
4                6.4        252         252  Phablet

In [16]:
elec.shape

(100, 4)

In [17]:
elec.Type.value_counts()

Tablet     50
Phablet    50
Name: Type, dtype: int64

In [29]:
tablet_price=elec[elec['Type']=='Tablet']['Price ($)']
phablet_price=elec[elec['Type']=='Phablet']['Price ($)']

In [30]:
print(stats.shapiro(tablet))
print(stats.shapiro(phablet))
print(stats.levene(tablet, phablet))

ShapiroResult(statistic=0.9741330742835999, pvalue=0.33798879384994507)
ShapiroResult(statistic=0.9686920642852783, pvalue=0.20465952157974243)
LeveneResult(statistic=0.6635459332943239, pvalue=0.4172859681296204)


In [ ]:
#Inference
#Data satisfies the condition of normaliry and the std of both the sample is same.
#therefore data satisfies the condition to use two sample z-test.

In [32]:
from statsmodels.stats import weightstats as stests 

In [48]:
test_stats, pvalue=stests.ztest(x1=tablet_price, x2=phablet_price,value=0,alternative='two-sided')

In [64]:
print('pvalue:', p_value)
print('test statistics', test_stats)

pvalue: 3.393300189145203e-21
test statistics 9.377016317496455


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>3. The technician claims that the price of phablet is not the same as the price of tablet. For the study, the technician collects the data of 50 gadgets of each type. Test the technician's claim using the 95% confidence interval.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Consider the dataset given in the CSV file `electronic.csv`. 

In [28]:
# type your code here
elec.head()

Screensize (inch)  Price ($)  Weight (g)     Type
0                6.5        245         252  Phablet
1                6.3        210         252  Phablet
2                6.1        224         234  Phablet
3                6.0        217         212  Phablet
4                6.4        252         252  Phablet

In [65]:
#Constructing the null hypothesis
#H0:Price of tablet and phablet is same
#H1: Price of tablet and phablet is not same
#It is two tailed test
tab_mean=np.mean(tablet_price)
phab_mean=np.mean(phablet_price)

n1=50
n2=50

tab_std=np.std(tablet_price)
phab_std=np.std(phablet_price)

null_hyp=0

#Finding the confidence interval
stats.norm.interval(0.95, loc=phab_mean-tab_mean, scale=np.sqrt(((tab_std)**2/n1)+((phab_std)**2/n2)))

#Since CI does not contain the 0, so reject the H0.


(35.87894938341398, 54.84105061658599)

In [69]:
#By p_value approach
num=phab_mean-tab_mean-null_hyp
den=((tab_std)**2/n1)+((phab_std)**2/n2)
test_stats=num/np.sqrt(den)
p_value=stats.norm.sf(test_stats)

In [67]:
test_stats

9.377016317496455

In [68]:
p_value

3.393300189145203e-21

In [59]:
#Critical value approach
z_critical=stats.norm.isf((1-0.95)/2)
z_critical

1.959963984540054

<a id="t"></a>
## 2. Small Sample Test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>4. The quality assurance department has collected 15 packets of potato chips. The department wants to test the average weight of the packets. Check whether they can use the one sample t-test for the population mean with 95% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

In [98]:
# type your code here
pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

In [99]:
len(pack_wt)

15

In [85]:
#H0: data is normal.
#H1: Data is not normal.
alpha=0.05
stats.shapiro(pack_wt)

ShapiroResult(statistic=0.9538426995277405, pvalue=0.5868644714355469)

Since pvalue is greater than the alpha=0.05,failed to reject the H0.

In [ ]:
#Since sample is drawn from the normally distributed population and n<30.
#We can use one sample t-test.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>5. Use the weight of the potato chips packets given in the previous question and test whether the average weight of packet is more than 30 g. using p-value method. (Use 10% level of significance)</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

In [4]:
pack_wt = [26.8, 29.6, 27.8, 31.2, 30.9, 27.1, 28, 28.6, 29.4, 29.3, 31.5, 32.4, 29.7, 28.1, 31.9]

import statistics
print(statistics.mean(pack_wt))

29.486666666666668


In [100]:
# type your code here
#Constructing the hypothesis
#H0: mu=30
#H1: mu>30
#It is right tailed test
mu=30
x_bar=np.mean(pack_wt)
sam_std=np.std(pack_wt)
n=len(pack_wt)
se=sam_std/np.sqrt(n)
test_stats=(x_bar-mu)/se
pvalue=stats.t.sf(test_stats, df=n-1)
print('Test Statistic:', test_stats)
print('p-value:', pvalue)

Test Statistic: -1.159799905970342
p-value: 0.8672358387939596


Inference
pvalue is gretaer than the alpha=0.1
Therefore failed to rejec t the H0.

In [97]:
stats.ttest_1samp(a=pack_wt, popmean=30, alternative='greater')

Ttest_1sampResult(statistic=-1.1204731591740873, pvalue=0.8593171187217965)

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>6. The orthopaedic surgeon states that a T-score for females older than 30 years is less than -1.2 which indicates the low bone density. To test the claim a sample of 10 women was selected and the average T-score was found to be -1.34 with a standard deviation of 0.8. Test the surgeon's claim using a critical value method with 95% confidence. (Assume the normality of the T-score). </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [107]:
# type your code here
#Constructing the hypothesis
# H0:T_score for female older than 30 year is -1.2
# H1:T_score for female older than 30 year is less than -1.2
#It is left tailed test
mu=-1.2
x_bar=-1.34
sam_std=0.8
n=10
#Finding the critical value
t_critical=-stats.t.isf(0.05, df=n-1)
#Finding the critical value
se=sam_std/np.sqrt(n)
t_stats=(x_bar-mu)/se
print('T critical value', t_critical)
print('Test Statistic:', t_stats)

T critical value -1.8331129326536337
Test Statistic: -0.5533985905294669


In [ ]:
#Since test stats is greater than the -1.83.
#Failed to reject the null hypothesis.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>7. The newspaper agency wants to test whether Harry takes less time to deliver the newspapers than his colleague Ron. The manager has collected time (in minutes) taken by Harry and Ron for 7 days. Is the given dataset unpaired? If yes, check whether the manager at newspaper agency can use the two sample t-test for unpaired data with 99% confidence.  </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    harry = [18.5, 17.4, 19.2, 16, 15.8, 13.4, 19.5]
    ron = [19.7, 18.6, 21.3, 17.5, 23.8, 20.7, 21]

In [108]:
# type your code here
#Yes the given dataset is unpaired.
harry = [18.5, 17.4, 19.2, 16, 15.8, 13.4, 19.5]
ron = [19.7, 18.6, 21.3, 17.5, 23.8, 20.7, 21]

In [109]:
#Use shapiro test
print(stats.shapiro(harry))
print(stats.shapiro(ron))
print(stats.levene(harry, ron))

ShapiroResult(statistic=0.9328558444976807, pvalue=0.575441837310791)
ShapiroResult(statistic=0.972867488861084, pvalue=0.9183047413825989)
LeveneResult(statistic=0.12793271359008374, pvalue=0.7267972480247995)


<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>8. The newspaper agency wants to test whether Harry takes less time to deliver the newspapers than his colleague Ron. The manager has collected time (in minutes) taken by Harry and Ron for 7 days. Use the given data and test the hypothesis for population mean time using a critical value method and p-value criteria with 90% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given below:
        
    harry = [18.5, 17.4, 19.2, 16, 15.8, 13.4, 19.5]
    ron = [19.7, 18.6, 21.3, 17.5, 23.8, 20.7, 21]

In [128]:
# type your code here
#Constructing the hypothesis
# H0:Harry and ron takes same time to deliver the newspapers.
# H1:Harry takes less time to deliver the newspapers than his colleague Ron.
#It is left tailed test
null=0
h_mean=np.mean(harry)
r_mean=np.mean(ron)
h_std=np.std(harry)
r_std=np.std(ron)
n1=len(harry)
n2=len(ron)

#Finding the critical value
t_crictical=-round(stats.t.isf(0.1,df=12),2)
print('T critical', t_crictical)
#Finding the test statistics and pValue
stats.ttest_ind(harry,ron)


T critical -1.36


Ttest_indResult(statistic=-2.8788582527177526, pvalue=0.01385995160611642)

Since,the test statistics is less than critical value, so by critical value approach we reject H0.
By p value approach,
pvalue is less than less tha 0.1.
Therefore reject H0.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                        <b>9. The pharmaceutical company had organized a program to introduce its new drug to lower the sugar level. They recorded the fasting sugar (in mg/dl) of 25 diabetic people. Those people undergo a medication of the new drug for 10 days and again took a fasting sugar level test. The company claim that the sugar level decreases due to its new drug. Test the claim using p-value technique with 95% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given in the file `sugar_level.xlsx`

In [124]:
# type your code here
sugar=pd.read_excel('sugar_level.xlsx')
sugar.head()

sugar_before  sugar_after
0           135          132
1           132          136
2           142          139
3           154          151
4           198          192

In [125]:
sugar.shape

(25, 2)

In [127]:
#Constructing the null hypothesis
# H0: mu1-mu2=0
# H1: mu1-mu2>0
# It is right tailed test
stats.ttest_rel(sugar['sugar_before'], sugar['sugar_after'])

Ttest_relResult(statistic=5.268469999584946, pvalue=2.108941394582287e-05)

Since the pvalue is less than 0.05.
So, reject the H0.
Therefore, medicine lowers the sugar level. 

<a id= "prop"></a>
## 3. Z Proportion Test

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>10. The physics department claims that it received 14% of fraud applications this year. The university head wants to test whether the percentage is different than what the department claims. The sample of 250 applications is selected out of which 38 applications are found to fraud. Test the claim with a 95% confidence interval.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [129]:
# type your code here
#H0: p=0.14
# H1: pIs not equal to 0.14
n=250
x=38
p_samp=x/n

stats.norm.interval(0.95,loc=p_samp, scale=np.sqrt((p_samp*(1-p_samp))/n))

(0.10749612708304186, 0.19650387291695814)

Since ci contains 0.14, we failed to reject H0.
Therefore department claim is true that 14% of fraud this year.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>11. The store manager claims that he received more 20% of defective plastic boxes in his previous order. Test the manager's claim using a critical value method. From a sample of 120 boxes, 42 were found to be defective. Test the claim with 90% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [135]:
# type your code here
# H0: p=0.20
#H1: p>0.20
#It is right tailed test

hyp_prop=0.2
sam_prop=42/120
n=120

#Finding the critical value
z_critical=stats.norm.isf(0.10)

test_stats=(sam_prop-hyp_prop)/np.sqrt((hyp_prop*(1-hyp_prop))/n)

print('Critical value', z_critical)
print('Test statistics', test_stats)

Critical value 1.2815515655446004
Test statistics 4.107919181288744


Since critical value is 1.28 and test stats is 4.10, which is more than the critical value.
So, we reject H0.
Therefore, Store manger claims is true that he receives more than 20% defective plastic boxes in his previous order.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                        <b>12. The HR head of a company claims that the company supports the woman empowerment and the proportion of female employees is same in the New York and Oneonta branches. The women empowerment cell wants to check whether the proportion is different for the two branches. The sample of 150 employees is selected from New York branch out of which 53 are females and a sample of 170 employees is selected from Oneonta branch out of which 76 are females. Use the p-value technique with 95% confidence. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [144]:
# type your code here
#H0: p1=p2
#H1: p1 is not equal to p2
#It is two tailed test.
ny_emp=150
on_emp=170


ny_female=53
on_female=76



In [145]:
import statsmodels.api as sm

In [146]:
sm.stats.proportions_ztest(count=np.array([ny_female, on_female]),nobs=np.array([ny_emp, on_emp]), alternative='two-sided')

(-1.705655272664087, 0.08807228185564836)

In [ ]:
#Pvalue is greater than the 0.05
#Failed to reject the H0.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>13. The education department claims that the proportion of students failed in Mathematics is more than the proportion of students failed in English. To test the claim a sample of 200 students enrolled for Mathematics was considered out of which 73 students failed in the exam and a sample of 150 students enrolled for English was considered out of which 53 students failed in the exam. Use the critical value method to test the department's claim with 90% confidence. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [152]:
# type your code here
#H0: proportion of students failed in Mathematics is same as the proportion of students failed in English
#H1: proportion of students failed in Mathematics is more than the proportion of students failed in English.
#It is right tailed test.
n_math=200
n_eng=150

fail_math=73
fail_eng=53
z_critical=stats.norm.isf(0.1)
print('critical value', z_critical)

critical value 1.2815515655446004


In [148]:
import statsmodels.api as sm

In [151]:
sm.stats.proportions_ztest(count=np.array([fail_math, fail_eng]), nobs=np.array([n_math, n_eng]),alternative='larger')

(0.22502571869471721, 0.41097963333926457)

Since, test statistic is 0.22 which is less than the critical value 1.28.
So, failed to reject the H0.
Therefore, proportion of students failed in Mathematics is not more than the proportion of students failed in English.